### PR TITLE
harden appscripts

### PR DESCRIPTION
**Pull request by szaimen**
fix https://github.com/nextcloud/vm/pull/1051#issuecomment-578543583

I updated all apps.

### CODE REVIEW

**Comment by szaimen** [comment]
@enoch85 I think this is ready for review.
Please check the PR-description.

File: apps/adminer.sh [hunk 1]
```diff
@@ -18,6 +16,15 @@ debug_mode
 # Check if root
 root_check
 
+# Check if adminer ist already installed
+if is_this_installed adminer
+then
+    msg_box "The 'adminer' script is currently not meant to run twice. Please create an issue in the Nextcloud/VM repository, if you would like to."
```
Comment (enoch85) [inline apps/adminer.sh]: We have a var called $ISSUES. ;)

**Comment by enoch85** [comment]
This is the easiest way of doing it, sure.

What I meant was making it possible to be run twice. Please work on that. I.e. Check what's installed, and it it's configured. Can it be overwritten? If yes, then don't care if it's installed twice. 

Test test test. :)

File: apps/adminer.sh (hunk 1 above)
Comment (szaimen) [reply apps/adminer.sh]: This should be done.

**Comment by szaimen** [comment]
@enoch85 The idea with the issue creation was, that I would pick it up if it was requested. Because I would like to do that in a follow up PR.

**Comment by enoch85** [comment]
> If an issue gets created, you can then just tag me and I will do it.

Problem is, users shouldn't have to do that, it should just work. 95% of the users won't create an issue even if needed, because they need to sign up to Github which is "naaah, I don't have time or energy, let's skip it".

**Comment by enoch85** [comment]
The slogan of Hansson IT is "with simplicity in mind". Expecting the users to create a PR because we are lazy not doing it correctly from start is not simple. :)

**Comment by szaimen** [comment]
> > If an issue gets created, you can then just tag me and I will do it.
> 
> Problem is, users shouldn't have to do that, it should just work. 95% of the users won't create an issue even if needed, because they need to sign up to Github which is "naaah, I don't have time or energy, let's skip it".

If thats true, then it was actually a really good idea 😅

**Comment by szaimen** [comment]
> The slogan of Hansson IT is "with simplicity in mind". Expecting the users to create a PR because we are lazy not doing it correctly from start is not simple. :)

okay, I will do what you want me to do, but it will take its time.

**Comment by szaimen** [comment]
@enoch85 Maybe we should discuss a bit more what should happen, if you execute an app a second time. Shall all settings get overwritten by default or should the settings stay, or what should happen?
I am not really sure. That was also the initial reason, why I didn't what to bother with this.

I also don't think, that any app has to be overwriteable. The reason, why I found the idea of making all apps available for everyone all the time, was that you could install an app later on, without having to tinker with it during the initial setup, because you have a main menu now all the time available, which is much more userfriendly imo.

File: apps/talk.sh [hunk 2]
```diff
@@ -20,6 +18,19 @@ debug_mode
 # Must be root
 root_check
 
+# Check if adminer ist already installed
+print_text_in_color "$ICyan" "Checking if Talk is already installed..."
+if is_this_installed coturn
```
Comment (enoch85) [inline apps/talk.sh]: Just because coturn is installed it doesn't mean that Talk is activated. Check with `occ` if any settings are made for example. 

[Here's](https://github.com/nextcloud/vm/blob/master/nextcloud_update.sh#L557-L577) and example of a good check.

Comment (szaimen) [reply apps/talk.sh]: Yes, thats right, but please answer my comment above first.
Afterwards I will improve the checks.

**Comment by enoch85** [comment]
> why I didn't want to bother with this.

You decided to make a PR which changes the behaviour of the scripts in total, making it easier/possible to run several times. This is just a second step of that PR. :)

Before the additional_apps PR you had to get the script from Github (the right one), and manage to run it. The barrier for most users are too steep, and I imagine that almost no one did it except the ones that know what they are doing. Yes, I get all the emails with questions regarding how to run things. :)

Imagine this, if new how to start a computer and run Windows on it, surf the web and that kind of stuff - what would you expect a script to to the second time you run it? A hazardous action like in overwriting your precious config files, or give you the option to do it, with a big warning, or merge the existing config into the current one? Making it easier for users to do things, also makes it more prune to doing mistakes. "What happens if I press this button"-kind-of-stuff. You get the point. :)

You don't have to over work it, Just make it very clear what's going on, and give the user the option to overwrite all current settings or something like that.

Sorry for not answering, I'm very busy right now.

**Comment by enoch85** [comment]
> okay, I will do what you want me to do, but it will take some time.

Totally fine with me, better doing it correct and 100% than 90% and not correct. :)

File: apps/adminer.sh [hunk 3]
```diff
@@ -16,14 +16,40 @@ debug_mode
...
+        check_external_ip
+        print_text_in_color "$ICyan" "Uninstalling Adminer and resetting all settings..."
+        a2disconf adminer.conf
+        rm $ADMINER_CONF
+        rm $ADMINERDIR/adminer.php
+        apt purge adminer -y
+        if ! restart_webserver
+        then
+        msg_box "Apache2 could not restart...
+        The script will exit."
```
Comment (enoch85) [inline apps/adminer.sh]: Maybe add $ISSUES here, or `check_command`? If not already implemented in `restart_server`

Comment (szaimen) [reply apps/adminer.sh]: I would say it is already implemented in restart_webserver:
https://github.com/nextcloud/vm/blob/40d9a0706e89a0180dc1fc1b4702a398c6bc541d/lib.sh#L414

Comment (szaimen) [reply apps/adminer.sh]: I would say it is already implemented in restart_webserver
https://github.com/nextcloud/vm/blob/40d9a0706e89a0180dc1fc1b4702a398c6bc541d/lib.sh#L414-L421

File: apps/talk.sh (hunk 2 above)
Comment (szaimen) [reply apps/talk.sh]: should be done now.

File: apps/collabora.sh [hunk 4]
```diff
@@ -23,6 +19,23 @@ root_check
...
+    msg_box "It seems like 'Onlyoffice' or 'Collabora' is already installed.\nIf you continue, Onlyoffice will be deleted and Collabora (re-)installed."
+    if [[ "no" == $(ask_yes_or_no "Do you really want to continue?") ]]
+    then
+        exit
+    fi
+fi
+
+print_text_in_color "$ICyan" "(Re-)Installing Collabora..."
+
+NC_UPDATE=1 && COLLABORA_INSTALL=1
```
Comment (enoch85) [inline apps/collabora.sh]: NC_UPDATE=1 && COLLABORA_INSTALL=1 . <(curl -sL https://raw.githubusercontent.com/nextcloud/vm/master/lib.sh)

It's stuff like this that makes me a bit worried. How can you miss the curl command? Aren't you testing your changes?

Comment (szaimen) [reply apps/collabora.sh]: I habe reverted that because i tested it and it didn't work ;)

Comment (enoch85) [reply apps/collabora.sh]: Haha ok! :D

**Comment by szaimen** [comment]
@enoch85 I think this is ready.
I have already tested everything tonight.
Please tell me, if you have some suggestions.
If you have nothing to add/change, I would test it one final one more this evening; but of course better would be, if you would test it as well to make sure everything works. (4 eyes principle) ;)
(some things are a bit hard to test, since I am running my nextcloud on my home net with NAT, so this is the only public ipaddress that I can currently test this on and would have to pause everything just to test this out, but I am doing my best.)

**Comment by enoch85** [comment]
@szaimen Good work here!

I can give you access to my Digitalocean account later so that you can test the scripts with a full install, that's how I do it. Or maybe you have your own VPS?

File: apps/collabora.sh [hunk 5]
```diff
@@ -23,6 +21,19 @@ root_check
 # Nextcloud 13 is required.
 lowest_compatible_nc 13
 
+# Check if onlyoffice ist already installed
```
Comment (enoch85) [inline apps/collabora.sh]: small typo here

File: apps/talk.sh [hunk 6]
```diff
@@ -20,6 +18,27 @@ debug_mode
...
+        exit
+    fi
+    print_text_in_color "$ICyan" "Uninstalling Talk and resetting all settings..."
+    occ_command_no_check config:app:delete spreed stun_servers
+    occ_command_no_check config:app:delete spreed turn_servers
+    occ_command_no_check app:remove spreed
+    rm $TURN_CONF
+    apt purge coturn -y
+    msg_box "Talk was successfully uninstalled and all settings were resetted."
+    exit
```
Comment (enoch85) [inline apps/talk.sh]: Do you exit the script here? Why?

Comment (szaimen) [reply apps/talk.sh]: Because I think it makes sense, if you run the script the first time and install it -> run it the second time and uninstall it -> run it the third time and reinstall it -> run it the fourth time and uninstall it again -> and so on.

Comment (szaimen) [reply apps/talk.sh]: That is how I imagine it to be for all scripts, but for some scripts it is currently too complicated to implement.

Comment (szaimen) [reply apps/talk.sh]: Because I think it makes sense, if you run the script the first time and install it -> run it the second time and uninstall it -> run it the third time and reinstall it -> run it the fourth time and uninstall it again -> and so on.

Comment (szaimen) [reply apps/talk.sh]: That is how I imagine it to be for all appscripts, but for some scripts it is currently too complicated to implement.

File: apps/collabora.sh (hunk 5 above)
Comment (szaimen) [reply apps/collabora.sh]: this is fixed

**Comment by szaimen** [comment]
> Or maybe you have your own VPS?

Currently not.

> I can give you access to my Digitalocean account later so that you can test the scripts with a full install.

That would be really nice!

File: apps/collabora.sh [hunk 7]
```diff
@@ -21,7 +21,7 @@ root_check
 # Nextcloud 13 is required.
 lowest_compatible_nc 13
 
-# Check if onlyoffice ist already installed
+# Check if collabora ist already installed
```
Comment (enoch85) [inline apps/collabora.sh]: ist/is (in all comments) :p

Comment (szaimen) [reply apps/collabora.sh]: oh fuck :D

Comment (szaimen) [reply apps/collabora.sh]: a generic copy and paste error ;)

Comment (szaimen) [reply apps/collabora.sh]: a generic copy and paste error ;)

Comment (szaimen) [reply apps/collabora.sh]: this is done.

**Comment by szaimen** [comment]
a first failure:
```
Sorry but something went wrong. Please report this issue to https://github.com/nextcloud/vm/issues and include the output of the error message. Thank you!
sudo -u www-data php /var/www/nextcloud/occ app:install spreedme failed
Posting notification to users that are admins, this might take a while...
```

Thats caused by spreedme which was the last time compatible with nc 13. So cannot test this out.
It just tries to reinstall it now with my script but has failed at the same point...

**Comment by szaimen** [comment]
btw: there is an error during the first installation of talk: (but that is not my fault)
`PHP Fatal error:  Cannot declare class OCA\Talk\Migration\Version2000Date20170707093535, because the name is already in use in /var/www/nextcloud/apps/spreed/lib/Migration/Version2000Date20170707093535.php on line 127
`

**Comment by szaimen** [comment]
@enoch85 I am ready with testing. All my changes seem to work as expected.
But I don't know how to fix travis...
If it is fixed, it could get merged imo.
Maybe the changes to spreedme could get reverted, but that is you decision.

**Comment by enoch85** [comment]
You can actually revert the checks for collabora and onlyoffice. They are already failsafe. Sorry for not mentioning that earlier, I've been super busy.

File: apps/adminer.sh [hunk 8]
```diff
@@ -18,16 +16,46 @@ debug_mode
...
+    msg_box "Adminer was successfully uninstalled and all settings were resetted."
+    exit
+fi
+
+print_text_in_color "$ICyan" "Installing and securing Adminer..."
+
+# Warn user about HTTP/2
+http2_warn Adminer
+
+check_external_ip
```
Comment (enoch85) [inline apps/adminer.sh]: Add comment?

File: apps/fail2ban.sh [hunk 9]
```diff
@@ -20,6 +18,26 @@ debug_mode
...
+    then
+        exit
+    fi
+    print_text_in_color "$ICyan" "Uninstalling Fail2Ban and resetting all settings..."
+    fail2ban-client unban --all
+    check_command apt purge fail2ban -y
+    rm /etc/fail2ban/filter.d/nextcloud.conf
+    rm /etc/fail2ban/jail.local
+    msg_box "Fail2Ban was successfully uninstalled and all settings were resetted."
+    exit
```
Comment (enoch85) [inline apps/fail2ban.sh]: Why does the script exit here when it could just continue with the installation?

File: apps/fulltextsearch.sh [hunk 10]
```diff
@@ -21,6 +19,19 @@ debug_mode
...
+if does_this_docker_exist "$nc_fts"
+then
+    msg_box "It seems like 'Fulltextsearch' is already installed.\nIf you continue, Fulltextsearch will get reinstalled."
+    if [[ "no" == $(ask_yes_or_no "Do you really want to continue?") ]]
+    then
+        exit
+    fi
+fi
+
+print_text_in_color "$ICyan" "(Re-)Installing Elastic Search & Full Text Search on Nextcloud..."
```
Comment (enoch85) [inline apps/fulltextsearch.sh]: Move reinstalling to the if statment and else Installing otherwise it will say Re-installing when it's the first time you install.

File: apps/spreedme.sh [hunk 11]
```diff
@@ -29,6 +27,19 @@ root_check
...
+if [ -d "$SNAPDIR" ]
+then
+    msg_box "It seems like 'Spreed.me' is already installed.\nIf you continue, it will be reinstalled."
+    if [[ "no" == $(ask_yes_or_no "Do you really want to continue?") ]]
+    then
+        exit
+    fi
+fi
+
+print_text_in_color "$ICyan" "(Re-)Installing Spreed.ME..."
```
Comment (enoch85) [inline apps/spreedme.sh]: Same thing here...

File: apps/talk.sh (hunk 6 above)
Comment (enoch85) [reply apps/talk.sh]: Hmm, I don't like that flow. 

Either it's installing for the first time, or it's reinstalling. For uninstall, please make a separate script, or a question, do you want to uninstall or re-install?

**Comment by enoch85** [comment]
Please change according to my comments in all scripts.

File: apps/talk.sh (hunk 6 above)
Comment (szaimen) [reply apps/talk.sh]: Okay, then another suggestion: after the uninstallation you can choose to directly install the app again. What do you think?

**Comment by szaimen** [comment]
@enoch85 do you have some thoughts one my comment concerning the installation-deinstallation flow?
https://github.com/nextcloud/vm/pull/1072#discussion_r372501145

File: apps/fail2ban.sh (hunk 9 above)
Comment (szaimen) [reply apps/fail2ban.sh]: because as I stated above I think it makes sense, if you can uninstall that app again. What about asking the user if he wants to continue to install the app again?

File: apps/talk.sh (hunk 6 above)
Comment (enoch85) [reply apps/talk.sh]: The flow should be:

1. New installation ---> install
2. Already installed ---> Warn about de-installing ----> de-install ----> ask if the user wants to reinstall---> re-install

File: apps/adminer.sh (hunk 8 above)
Comment (enoch85) [reply apps/adminer.sh]: ping @szaimen Why isn't check_external_ip in lib.sh, and are there more occurrences like this that would need to be replaced?

File: apps/talk.sh (hunk 6 above)
Comment (enoch85) [reply apps/talk.sh]: @szaimen When you have time, please take a look at this.

File: apps/adminer.sh (hunk 8 above)
Comment (enoch85) [reply apps/adminer.sh]: @szaimen Can you answer this please?

File: apps/collabora.sh [hunk 12]
```diff
@@ -5,11 +5,7 @@
 # shellcheck disable=2034,2059
 true
```
Comment (enoch85) [inline apps/collabora.sh]: no need to change collabora, or onlyoffice.

**Comment by enoch85** [comment]
When you have time @szaimen, please take a look at my comments and fix accordingly.

No need for change:
- Netdata
- OnlyOffice
- Collabora

**Comment by szaimen** [comment]
> No need for change:
> 
> * Netdata
> * OnlyOffice
> * Collabora

Yes, maybe they are already failproof, but I would like to jnform the user, when the app is already installed and let the user choose if he wants to continue.

If there is also a way to remove the whole app, I will provide a radiolist, so that the user can choose what he wants to do (e.g. uninstall or reinstall).

**Comment by enoch85** [comment]
> If there is also a way to remove the whole app

https://github.com/nextcloud/vm/blob/master/apps/collabora.sh#L74-L78

File: apps/adminer.sh (hunk 8 above)
Comment (szaimen) [reply apps/adminer.sh]: It was before a script, that I have made a function from, because I need it two times in the script. I will put it into the lib.sh

Comment (szaimen) [reply apps/adminer.sh]: and no, there are not more occurentes like this afaik.

Comment (szaimen) [reply apps/adminer.sh]: this should be done.

File: apps/fulltextsearch.sh (hunk 10 above)
Comment (szaimen) [reply apps/fulltextsearch.sh]: this should be done 👍

File: apps/spreedme.sh (hunk 11 above)
Comment (szaimen) [reply apps/spreedme.sh]: this also should be done

File: apps/collabora.sh (hunk 12 above)
Comment (szaimen) [reply apps/collabora.sh]: what shall I do here? change this to `false`?

Comment (szaimen) [reply apps/collabora.sh]: Or did you meant changing those scripts at all?

File: apps/talk.sh (hunk 6 above)
Comment (szaimen) [reply apps/talk.sh]: this should be done.

File: apps/fail2ban.sh (hunk 9 above)
Comment (szaimen) [reply apps/fail2ban.sh]: this should be done.

**Comment by szaimen** [comment]
> > If there is also a way to remove the whole app
> 
> https://github.com/nextcloud/vm/blob/master/apps/collabora.sh#L74-L78

If collabora, onlyoffice or fulltextsearch shall be removed, I think it is not enough to remove the nextcloud app. But I am also not completely sure, if it is enough to just remove the docker container, since you are also making changes to the reverse proxy, etc. That is the reason why these applications aren't able to get uninstalled currently.

**Comment by szaimen** [comment]
@enoch85 do you know how to fix travis?

File: apps/collabora.sh (hunk 12 above)
Comment (enoch85) [reply apps/collabora.sh]: > Or did you meant changing those scripts at all?

Yes, I think the function already is good as is.

**Comment by enoch85** [comment]
> But I am also not completely sure

I'm sure how it works, and it works in the current form, so please don't change it. :)

**Comment by enoch85** [comment]
> @enoch85 do you know how to fix travis?

Yes, remove your changes to Collabora and OnlyOffice + Netdata as I mentioned before. I will reinstall already if already installed.

**Comment by szaimen** [comment]
> > @enoch85 do you know how to fix travis?
> 
> Yes, remove your changes to Collabora and OnlyOffice + Netdata as I mentioned before. I will reinstall already if already installed.

The thing is like I mentioned above, that you don't get the chance to abort the execution, it directly starts the reinstallation (aka "please enter the domain" and so on). I would like to make this cleaner and this is the way I think it works best from a UX point of view.

Also later on you could add a step to uninstall it. (remove the docker container, reset the nextcloud-app-settings and uninstall the nextcloud-app)

Also netdata can get deinstalled fairly easy and this is what I wanted to provide here. (but the current deinstallation process for netdata hasn't worked without my changes in my testing btw...)

**Comment by enoch85** [comment]
Actually, I tried the apps.sh yesterday and explained it to a customer. When hearing myself say it out loud, it's actually really nice.

I would like to prioritize this so that we can release 18.0.1 soon.

Regarding [this](https://github.com/nextcloud/vm/pull/1072#issuecomment-585751899) do what you think works, and I'll think about it.

I'm just saying that you wouldn't run the app script if you didn't wanted to install, right? Hence the questions comes up first (Collabora & OnlyOffice) and depending on if any of them are installed, it's automatically removed/re-installed. No need to ask the user that.

Regarding Netdata, we do reinstall every time we run the update script, and it never failed so far: https://github.com/nextcloud/vm/blob/master/nextcloud_update.sh#L97-L107

The thing is, it also needs to be backwards compatible, since Netdata decided to change the path of the script in later versions. A lot to think about, but stability comes first on my agenda. :)

**Comment by szaimen** [comment]
> Actually, I tried the apps.sh yesterday and explained it to a customer. When hearing myself say it out loud, it's actually really nice.

Nice to here that, I like it, too! :D

I will think through all points you have mentioned tomorrow and will come up with a solution. 👍

**Comment by szaimen** [comment]
> Regarding [this](https://github.com/nextcloud/vm/pull/1072#issuecomment-585751899), do what you think works, and I'll think about it.

I have done what I think works best now.

> I'm just saying that you wouldn't run the app script if you didn't wanted to install, right? Hence the questions comes up first (Collabora & OnlyOffice) and depending on if any of them are installed, it's automatically removed/re-installed. No need to ask the user that.

I have made options available to uninstall and reinstall collabora, onlyoffice and fulltextsearch.

> Regarding Netdata, we do reinstall every time we run the update script, and it never failed so far: https://github.com/nextcloud/vm/blob/master/nextcloud_update.sh#L97-L107
> 
> The thing is, it also needs to be backwards compatible, since Netdata decided to change the path of the script in later versions. A lot to think about, but stability comes first on my agenda. :)

Netdata should now be resolved by executing the update script directly over the update script.

**Comment by szaimen** [comment]
@enoch85 I think this is ready for review now. Please tell me what you think about the solutions that I have found. 

And BTW: I still don't know how to fix travis. Do you maybe have an idea?

File: apps/collabora.sh [hunk 13]
```diff
@@ -80,6 +80,21 @@ install_docker
...
+    occ_command app:remove documentserver_community
+fi
+
+# Disable RichDocuments (Collabora App) if activated
+if [ -d "$NC_APPS_PATH"/richdocuments ]
+then
+    occ_command app:remove richdocuments
+fi
+
 # Disable RichDocuments (Collabora App) if activated
```
Comment (enoch85) [inline apps/collabora.sh]: disable it two times?

**Comment by enoch85** [comment]
This PR is getting messy, please make a new one without any double code, and clean code please.

Thanks <3

**Comment by szaimen** [comment]
Do you maybe have an idea how we could prevent double code here? Because I think it is needed here. Maybe we could make a function from the double code instead...

**Comment by enoch85** [comment]
> disable it two times?

I think that is messy for example. 

Basically, if you run an install script- what is you intention? Installing the app, right? So why all these questions here and there? It just makes it messy if you ask me. 

Sure, add an uninstall menu like you did in the Onlyoffice script with two options: 
1 Uninstall 
2. Reinstall. 

The scripts should be that good that the users doesn't need to interact with them. We should fix that script wise. A good example is the Collabora (and OnlyOffice) script:

1. The user has OnlyOffice installed, but that doesn't work for some reason or wants to change
2. The user run the installl script of Collabora
3. We remove all previous versions of both Collabora *and* Onlyoffice (since they are not compatible with each other). We assume that's the goal here and use `any_key` with the option to abort if that's not the intention.
4. The user gets Collbora by just hitting enter, enter, enter (next, next, next). No need to interact really.

Sure, we inform the user about what's going on, but we assume what he/she wants - meaning, a much better (easier) flow instead of a big "ATTENTION you will now uninstall the app" (the user would feel that I guess, I know how I would react as a inexperienced Linux user) when it doesn't really matter, because you're not loosing any data, just reinstalling a function (app).

You see the picture here?

**Comment by szaimen** [comment]
> > disable it two times?
> 
> I think that is messy for example.

Actually, I think, that you haven't checked out the current state. Please do that. There shouldn't be richdocuments that get disabled twice to my knowledge.

> Basically, if you run an install script- what is you intention? Installing the app, right? So why all these questions here and there? It just makes it messy if you ask me.
> 
> Sure, add an uninstall menu like you did in the Onlyoffice script with two options:
> 1 Uninstall (remove all settings associated with it)
> 2. Reinstall (overwrite all settings associated with new settings and remove legacy)
> 

that is exacly what this PR has done.

> The scripts should be that good that the users doesn't need to interact with them. We should fix that script wise. A good example is the Collabora (and OnlyOffice) script:
> 
> 1. The user has OnlyOffice installed, but that doesn't work for some reason or wants to change
> 2. The user run the installl script of Collabora
> 3. We remove all previous versions of both Collabora _and_ Onlyoffice (since they are not compatible with each other). We assume that's the goal here and use `any_key` with the option to abort if that's not the intention.
> 4. The user gets Collbora by just hitting enter, enter, enter (next, next, next). No need to interact really.

So what do you want then, aks the user if he wants to uninstall or reinstall the app in the first place with a whiptail or just any_key to continue deinstalling the app, with a hint, that you get the change to install it again afterwards?

> Sure, we inform the user about what's going on, but we assume what he/she wants - meaning, a much better (easier) flow instead of a big "ATTENTION you will now uninstall the app" (the user would feel that I guess, I know how I would react as a inexperienced Linux user) when it doesn't really matter, because you're not loosing any data, just reinstalling a function (app).
> 
> You see the picture here?

The big attention doesn't have to be that big... but that would come with any_key imho.

File: apps/collabora.sh (hunk 13 above)
Comment (szaimen) [reply apps/collabora.sh]: that isn't the current state of this PR
Don't know why you have checked that out?

**Comment by szaimen** [comment]
@enoch85 any thoughts on my comment above? https://github.com/nextcloud/vm/pull/1072#issuecomment-586692706
I would like to finish this finally.

**Comment by enoch85** [comment]
I'll have a look in a moment.

**Comment by enoch85** [comment]
@szaimen Have you tested all the scripts individually?

File: apps/collabora.sh [hunk 14]
```diff
@@ -23,10 +19,78 @@ root_check
...
+# If yes, then stop and prune the docker container
+docker_prune_this 'onlyoffice/documentserver'
+
+# remove OnlyOffice-documentserver if activated
+install_if_not jq
+if occ_command_no_check app:list --output=json | jq -e '.enabled | .documentserver_community' > /dev/null
+then
+    msg_box "Removing OnlyOffice-documentserver: $1\nYou will be given the option to abort when you hit OK."
+    any_key "Press any key to continue. Press CTRL+C to abort"
+    occ_command app:remove documentserver_community
```
Comment (enoch85) [inline apps/collabora.sh]: This logic isn't needed reallt as `occ_command app:remove documentserver_community` will remove it if it's installed, or just say that there's nothing to remove as it's not installed anyway.

Please don't add stuff that might break later on. I mean, what if the text you are grepping for are changed later on? Then the check would fail anyway, and `occ app:remve` alrady have a check built in.

That's what I meant before.

**Comment by enoch85** [comment]
I will test this when you have tested it yourself. 

There are some errors in these changes that you would catch if you tested it yourself.

File: apps/adminer.sh [hunk 15]
```diff
@@ -7,8 +7,6 @@ true
 # shellcheck source=lib.sh
 . <(curl -sL https://raw.githubusercontent.com/nextcloud/vm/master/lib.sh)
 
-print_text_in_color "$ICyan" "Installing and securing Adminer..."
```
Comment (enoch85) [inline apps/adminer.sh]: There's a reason why this comes first in all scrips. If you run it on a slow connection then it will take time to load the lib.sh and the user might think the setup-script has stopped.

Please test this branch by using the setup script and choose all apps. Then you'll see what I mean. Even with a fast connection there's a lag of about 3 seconds. If we can tell the user what's going on before he/she starts to worry, that's a good thing.

File: apps/adminer.sh [hunk 16]
```diff
@@ -18,16 +16,52 @@ debug_mode
...
+            # Check that the script can see the external IP (apache fails otherwise)
+            check_external_ip
+            print_text_in_color "$ICyan" "Uninstalling Adminer and resetting all settings..."
+            a2disconf adminer.conf
+            rm $ADMINER_CONF
+            rm $ADMINERDIR/adminer.php
+            check_command apt purge adminer -y
+            if ! restart_webserver
+            then
+                msg_box "Apache2 could not restart...\nThe script will exit."
```
Comment (enoch85) [inline apps/adminer.sh]: If Apache stops, that's an hard error caused by this script. Better report it to $ISSUES so that we can fix it.

File: apps/collabora.sh [hunk 17]
```diff
@@ -5,11 +5,7 @@
 # shellcheck disable=2034,2059
 true
 # shellcheck source=lib.sh
```
Comment (enoch85) [inline apps/collabora.sh]: Here's how it should look like. This will fix your travis issues.

File: apps/collabora.sh [hunk 18]
```diff
@@ -23,10 +19,78 @@ root_check
...
+
+# Check if OnlyOffice is previously installed
+# If yes, then stop and prune the docker container
+docker_prune_this 'onlyoffice/documentserver'
+
+# remove OnlyOffice-documentserver if activated
+install_if_not jq
+if occ_command_no_check app:list --output=json | jq -e '.enabled | .documentserver_community' > /dev/null
+then
+    msg_box "Removing OnlyOffice-documentserver: $1\nYou will be given the option to abort when you hit OK."
```
Comment (enoch85) [inline apps/collabora.sh]: To noisy if you ask me. Just let the user know that Onlyoffice will be uninstalled with a countdown, and the option to abort with CTRL+C. It's always possible to reinstall it later if that's the intention. 

This just takes time, and is un-smooth.

File: apps/fulltextsearch.sh [hunk 19]
```diff
@@ -6,11 +6,7 @@
 # shellcheck disable=2034,2059
 true
 # shellcheck source=lib.sh
-NC_UPDATE=1 && ES_INSTALL=1 . <(curl -sL https://raw.githubusercontent.com/nextcloud/vm/master/lib.sh)
-unset NC_UPDATE
-unset ES_INSTALL
-
-print_text_in_color "$ICyan" "Installing Elastic Search & Full Text Search on Nextcloud..."
+. <(curl -sL https://raw.githubusercontent.com/nextcloud/vm/master/lib.sh)
```
Comment (enoch85) [inline apps/fulltextsearch.sh]: Why did you remove `NC_UPDATE=1 && ES_INSTALL=1` from here? Do you know the impact of removing it?

File: apps/onlyoffice.sh [hunk 20]
```diff
@@ -74,23 +139,6 @@ check_open_port 443 "$SUBDOMAIN"
...
-fi
-
-# Disable OnlyOffice (Collabora App) if activated
-if [ -d "$NC_APPS_PATH"/onlyoffice ]
-then
-    occ_command app:remove onlyoffice
-fi
-
 # Install Onlyoffice docker
 docker pull onlyoffice/documentserver:latest
```
Comment (enoch85) [inline apps/onlyoffice.sh]: You can remove OnlyOffice from this PR as we will do that separately.

File: apps/spreedme.sh (hunk 11 above)
Comment (enoch85) [inline apps/spreedme.sh]: Don't install spreed.ME, install Talk instead. Spreed.ME isn't even in the startup script anymore.

File: apps/tmbitwarden.sh [hunk 21]
```diff
@@ -18,6 +17,16 @@ debug_mode
...
+then
+    msg_box "It seems like 'Bitwarden' is already installed.\nYou cannot run this script twice, because you would loose all your passwords."
+    exit 1
+fi
+
+print_text_in_color "$ICyan" "Installing Bitwarden password manager..."
+
 msg_box "Bitwarden is a password manager that is seperate from Nextcloud, though we provide this service because it's self hosted and secure.
 
 If you just want to run Bitwarden locally (not connecting your smartphone) then you can use 'localhost' as domain.
```
Comment (enoch85) [inline apps/tmbitwarden.sh]: :+1:

Comment (enoch85) [reply apps/tmbitwarden.sh]: But, use the checks from the update script to really check if Bitwarden is installed or not: https://github.com/nextcloud/vm/blob/master/nextcloud_update.sh#L557-L577

File: apps/webmin.sh [hunk 22]
```diff
@@ -18,6 +16,32 @@ debug_mode
...
+then
+    choice=$(whiptail --radiolist "It seems like 'Webmin' is already installed.\nChoose what you want to do.\nSelect by pressing the spacebar and ENTER" "$WT_HEIGHT" "$WT_WIDTH" 4 \
+    "Uninstall Webmin" "" ON \
+    "Reinstall Webmin" "" OFF 3>&1 1>&2 2>&3)
+    
+    case "$choice" in
+        "Uninstall Webmin")
+            print_text_in_color "$ICyan" "Uninstalling Webmin..."
+            check_command apt purge webmin -y
+            msg_box "Webmin was successfully uninstalled."
```
Comment (enoch85) [inline apps/webmin.sh]: add `apt purge autoremove -y` to remove dependencies as well

**Comment by szaimen** [comment]
Tanks for reviewing. I will resolve all points soon!

File: apps/fulltextsearch.sh (hunk 19 above)
Comment (szaimen) [reply apps/fulltextsearch.sh]: It is moved to a later point in the script.

File: apps/spreedme.sh (hunk 11 above)
Comment (szaimen) [reply apps/spreedme.sh]: So shall I remove all changed that I have made to spreed.me?

File: apps/onlyoffice.sh (hunk 20 above)
Comment (szaimen) [reply apps/onlyoffice.sh]: Okay 👍

File: apps/collabora.sh (hunk 17 above)
Comment (szaimen) [reply apps/collabora.sh]: So shall I add this before every line of the travis error lines? If I remember correctly, than I have already tried that but then a different travis error appeared.

File: apps/collabora.sh (hunk 18 above)
Comment (szaimen) [reply apps/collabora.sh]: Okay, you are right.

File: apps/tmbitwarden.sh (hunk 21 above)
Comment (szaimen) [reply apps/tmbitwarden.sh]: Okay

File: apps/webmin.sh (hunk 22 above)
Comment (szaimen) [reply apps/webmin.sh]: Okay

File: apps/adminer.sh (hunk 16 above)
Comment (szaimen) [reply apps/adminer.sh]: Okay

File: apps/adminer.sh (hunk 15 above)
Comment (szaimen) [reply apps/adminer.sh]: Sounds reasonable. But I will change the lines to e.g.: "Downloading and executing Collabora.sh'

File: apps/collabora.sh (hunk 14 above)
Comment (szaimen) [reply apps/collabora.sh]: The check was for canceling the uninstallation of onlyoffice. But if you like it more this way, I will do so.

File: apps/fulltextsearch.sh (hunk 19 above)
Comment (enoch85) [reply apps/fulltextsearch.sh]: OK, for what reason?

**Comment by enoch85** [comment]
121 commits
104 conversations
over 400 changes

It's better if we do this in new PRs, one for each script.

**Comment by szaimen** [comment]
Makes sense 👍
I will split them up later on.

File: apps/fulltextsearch.sh (hunk 19 above)
Comment (szaimen) [reply apps/fulltextsearch.sh]: There is no good reason. I will revert that change.

File: apps/adminer.sh (hunk 16 above)
Comment (szaimen) [reply apps/adminer.sh]: this is done.

**Comment by szaimen** [comment]
I will first fix all reviews and split them up in seperate PRs afterwards.

File: apps/collabora.sh (hunk 14 above)
Comment (szaimen) [reply apps/collabora.sh]: this is done.

File: apps/collabora.sh (hunk 17 above)
Comment (szaimen) [reply apps/collabora.sh]: I hope that works. will see later on.

File: apps/collabora.sh (hunk 18 above)
Comment (szaimen) [reply apps/collabora.sh]: this should be done.

File: apps/fulltextsearch.sh (hunk 19 above)
Comment (szaimen) [reply apps/fulltextsearch.sh]: is reverted.

File: apps/spreedme.sh (hunk 11 above)
Comment (szaimen) [reply apps/spreedme.sh]: all changes to spreed.me have been reverted.

File: apps/tmbitwarden.sh (hunk 21 above)
Comment (szaimen) [reply apps/tmbitwarden.sh]: this should be done.

File: apps/webmin.sh (hunk 22 above)
Comment (szaimen) [reply apps/webmin.sh]: this should be done.

File: apps/adminer.sh (hunk 15 above)
Comment (szaimen) [reply apps/adminer.sh]: this is fixed.

**Comment by szaimen** [comment]
I think I have fixed all reviews now and will handle the splitup like this:
I will copy the changes to a new PR for each app and mark them as WIP. 
When I have tested the PR, I will mark it as ready. (disable the WIP-status)

**Comment by szaimen** [comment]
I think all changes are copied over.
